[Mono] Use dn_vector_t instead of GArray in a few places

lambdageek opened the pull request:
Set up the infrastructure to consume shared container types from src/mono/{utils,metadata,containers,mini}.

There are about 20 total places where we use GArray. Replace a few of them with dn_vector_t as an experiment to see how using shared containers in Mono will look like.

lambdageek committed:
Use dn_vector_t instead of GArray in one place in Mono

Set up the infrastructure to consume shared container types from
src/mono/{metadata,mini}.

Replace one use of GArray with dn_vector_t

lambdageek commented on src/mono/mono/metadata/CMakeLists.txt:
```diff
@@ -1,5 +1,19 @@
 project(metadata C)
 
+include (${SHARED_CONTAINERS_SOURCE_PATH}containers.cmake)
+
+set(shared_container_sources "")
+
+list(APPEND shared_container_sources
+  ${SHARED_CONTAINER_SOURCES}
+  ${SHARED_CONTAINER_HEADERS}
+)
+
+addprefix(shared_container_sources ${SHARED_CONTAINERS_SOURCE_PATH} "${shared_container_sources}")
+
+add_library(shared_container_objects OBJECT ${shared_container_sources})
+target_include_directories(shared_container_objects PUBLIC ${SHARED_CONTAINERS_SOURCE_PATH})
+
```
this should probably be done in `src/mono/mono/CMakeLists.txt`, not in metadata

lambdageek replied:
Turns out this isn't a nice to have. it's necessary in order to avoid duplicate symbols in the static linking case.  additionally I need to tackle eventpipe's use of the shared containers in this PR, not separately.

lambdageek commented on src/mono/mono/metadata/custom-attrs.c:
```diff
@@ -1904,7 +1906,10 @@ mono_custom_attrs_from_index_checked (MonoImage *image, guint32 idx, gboolean ig
-	attr_array = g_array_sized_new (TRUE, TRUE, sizeof (guint32), 128);
+	dn_vector_custom_alloc_params_t vec_params = {0,};
+	vec_params.capacity = 128;
+	vec_params.attributes = DN_VECTOR_ATTRIBUTE_MEMORY_INIT;
```
Not really necessary since we don't have any gaps in the data in this array and we never look past the end.

lambdageek committed:
include shared containers in mono shared lib, too

lambdageek committed:
convert mono_seq_point_init_next to use dn_vector_t

lambdageek committed:
[custom-attrs] Use init/dispose instead of alloc/free for dn_vector_t

Use a stack allocated container (with heap-allocated data) to avoid
one extra allocation

lambdageek committed:
move container logic to toplevel; add deps to runtime components

Add a way to specify target_link_libraries for runtime components.
Use it to link in the shared containers.

Avoids duplicate linking of the container objects in static linking scenarios.

lambdageek committed:
fix windows build

lambdageek committed:
replace 2 uses of GArray in interp/transform.c

lambdageek committed:
replace a use of GArray in mini/seq-points.c

lambdageek committed:
replace use of GArray for debug-mini

lambdageek committed:
fix Darwin frameworks builds

lambdageek commented:
This is just a start.  There are around 20 uses of `g_array_new...` in Mono, 89 `g_ptr_array_new...`, 7 `g_byte_array_new`.

And after that there's 289 `g_hash_table_new...` 😱

BrzVlad commented:
Is allocation from these different ? I think all our g_* data types were using malloc behind the scenes.

lambdageek commented:
> Is allocation from these different ? I think all our g_* data types were using malloc behind the scenes.

by default they use malloc/free. you can customize the containers to use custom allocators (right now Johan added "allocate from a fixed buffer" and "allocate from a fixed buffer, overflow to malloc/free").  I was thinking of eventually adding a mempool allocator, although we might need to tweak the allocator interfaces a little bit.

lateralusX commented on src/mono/mono/mini/interp/transform.c:
```diff
@@ -3939,7 +3940,8 @@ recursively_make_pred_seq_points (TransformData *td, InterpBasicBlock *bb)
 
-	GArray *predecessors = g_array_new (FALSE, TRUE, sizeof (gpointer));
+	dn_vector_t predecessors = {0,};
+	dn_vector_init_t (&predecessors, SeqPoint*);
```
An alternative would be to use dn_vector_ptr_t for this, that will have a little more optimized push_back/pop_back implementations.

lambdageek replied:
Can I use a `dn_vector_ptr_t` with `dn_vector_index_t`?  I think the answer is yes....

lateralusX replied:
Yes, there is also a dn_vector_ptr_index implemented as:

return dn_vector_index_t ((dn_vector_t *)vector, type, index);

so it will cast to type.

lateralusX replied:
Maybe we should add a couple of _t versions to ptr as well? That can be done in dn-vector-ptr.h, like:

dn_vector_ptr_index_t
dn_vector_ptr_at_t
dn_vector_ptr_front_t
dn_vector_ptr_back_t
dn_vector_ptr_data_t

That way you can use the generic dn_vecotr_ptr_t type and still open up to cast to different pointer types when accessing the elements.

lambdageek replied:
Added

lateralusX commented on src/mono/mono/mini/debug-mini.c:
```diff
@@ -78,7 +79,7 @@ mono_debug_open_method (MonoCompile *cfg)
 
 	info->jit = jit = g_new0 (MonoDebugMethodJitInfo, 1);
-	info->line_numbers = g_array_new (FALSE, TRUE, sizeof (MonoDebugLineNumberEntry));
+	info->line_numbers = dn_vector_alloc_t (MonoDebugLineNumberEntry);
```
dn_vector_alloc_t can fail and return NULL.

lambdageek committed:
WIP: start a dn_checkfail helper like g_assertf

Will need the same kind of output redirection that mono does for
Android and iOS.

lambdageek committed:
remove some unused assignments from eventpipe.cmake fragment

We get the shared containers via a dependency on the containers object
library now

lambdageek committed:
make containers a normal sub-project for Mono

Preliminary step before adding runtime-specific container code

lambdageek committed:
Add mono implementation of dn-rt.h

lambdageek committed:
check dn_vector_alloc_t result

lambdageek committed:
fix coreclr build

lambdageek commented:
/azp run runtime

lambdageek committed:
add placeholder dn_rt functions for coreclr and nativeaot

lambdageek committed:
fix mono build

lambdageek committed:
maybe fix coreclr/nativeaot windows build?

lambdageek committed:
move nativeaot shims to a separate file

fix windows include problems, I guess?

lambdageek committed:
fix coreclr windows?

lambdageek committed:
add typed dn_vector_ptr accessor macros

lambdageek committed:
Use dn_vector_ptr in a few places instead of dn_vector

lambdageek commented:
@LakshanF could you take a look at the nativeaot (and coreclr?) bit.  There's just one thing here:

I'm adding a checked fast fail macro `dn_checkfail(cond,msg,...)` that checks if `cond` is true and if not, fast fails (in both release and debug builds).  It's not actually used inside the containers library anywhere, but I want to start using it in Mono as we migrate to the shared containers to preserve our old behavior (unlike asserts in coreclr, mono's `g_assert` macro fires in both debug and release builds).

The nativeaot version just ignores the message for now and calls `RhFastFail`.  I could try to make it print out the message, but I wasn't sure if there's some Redhawk way of doing that or what the general policy in native aot is around stuff like that.

lateralusX commented on src/mono/mono/containers/dn-rt-utils.c:
```diff
@@ -0,0 +1,19 @@
+// Licensed to the .NET Foundation under one or more agreements.
```
Shouldn't the naming of this file follow the naming of the header, dn-rt-mono.c?

lateralusX commented on src/mono/mono/mini/debug-mini.c:
```diff
@@ -78,7 +79,8 @@ mono_debug_open_method (MonoCompile *cfg)
 	info->jit = jit = g_new0 (MonoDebugMethodJitInfo, 1);
-	info->line_numbers = g_array_new (FALSE, TRUE, sizeof (MonoDebugLineNumberEntry));
+	info->line_numbers = dn_vector_alloc_t (MonoDebugLineNumberEntry);
+	dn_checkfail (info->line_numbers != NULL, "Allocation failed");
```
Thinking a little about this pattern, could an alternative be to have a new vector attribute that we can set when doing custom alloc/init that we fail fast internal, like DN_VECTOR_ATTRIBUTE_FAIL_FAST. That code path is already in the slow path so check this attribute and fail fast if requested instead of returning NULL should not add overhead to fast path. It will be an opt in and you need to use the custom init/alloc to get it, but will reduce need to check returns due to allocation failures, that could happen to a number of different functions, alloc/init/push_back/resize etc.

lateralusX replied:
We could also add a couple of pre defined init structs that enable some values, so instead of needing to declare a struct with this flag set, you can just use the predeclared struct in your custom init/alloc call.

lambdageek replied:
Honestly that all sounds like a huge pain.

I would prefer something like

```
  dn_vector_t *vec = DN_AOK (dn_vector_alloc_t (...));
```

with something like

```c
#define DN_AOK(expr) ({ void *_ck = (expr); dn_checkfail(_ck != NULL, "Allocation failed"); _ck})
```

except:
1. with proper macros so the file:line is from the caller of `DN_AOK()`, not from the  macro definiton
2. with some `typeof` trick so the whole expression has the same type as `expr`
3. without using the GCC `({ stmt; expr })` non-standard extension

lateralusX replied:
I think it could be rather straightforward, and it would simplify case to work similar to how g_array works where it asserts on internal allocation failures, so you could make sure a dn_vector_t would behave the same way with regards to internal allocation failures.

The macro way works as well, but they are not exclusive, both variations could exist. Handle it using macros will of course generate more code around each call to a vector function that could fail (that we didn't check for failures when using g_array). The macro solution would probably be able to validate both pointer as well as bool return cases for failures.

lateralusX commented on src/mono/mono/mini/interp/transform.c:
```diff
@@ -3939,7 +3941,8 @@ recursively_make_pred_seq_points (TransformData *td, InterpBasicBlock *bb)
 	SeqPoint ** const MONO_SEQ_SEEN_LOOP = (SeqPoint**)GINT_TO_POINTER(-1);
 
-	GArray *predecessors = g_array_new (FALSE, TRUE, sizeof (gpointer));
+	dn_vector_ptr_t predecessors = {0,};
```
No need to init the dn_vector_ptr_t, it will be memset by dn_vector_ptr_init.

lambdageek replied:
I'm not sure if C compilers (or linters) would like that.

lateralusX commented on src/mono/mono/mini/seq-points.c:
```diff
@@ -33,7 +35,9 @@ recursively_make_pred_seq_points (MonoCompile *cfg, MonoBasicBlock *bb)
 
-	GArray *predecessors = g_array_new (FALSE, TRUE, sizeof (gpointer));
+
+	dn_vector_ptr_t predecessors = {0,};
```
No need to init the dn_vector_ptr_t, it will be memset by dn_vector_ptr_init.

lambdageek commented:
Just to give a brief update: I'm probably going to pause on this until Preview 4 is out.

Also in the follow-up branch where I try to get rid of all uses of `GPtrArray` I found a few more cmake tweaks that I need to make.   I will bring those over to this PR, too, so we have all the build stuff sorted out.

lambdageek commented:
@lateralusX I'll get back to this next week.   hope I can at least get the basic infrastructure in place
